scanners/third_parties.js

```diff
@@ -0,0 +1,151 @@
+
+  // Isolate the hostname with or without a www prefix,
+  // and treat them effectively as the same hostname.
+  if (url.hostname.startsWith("www.")) {
+    www_host = url.hostname
+    root_host = www_host.replace(/^www\./, "")
+  } else {
+    www_host = "www." + url.hostname
+    root_host = url.hostname
+  }
+
+  base_host = baseDomainFor(root_host);
+  source_base = baseDomainFor(source.hostname);
+
+  /***
+  * There are 4 cases:
+  * - internal: same hostname (or with a www prefix) as the source URL.
+  * - nearby: same base domain, different (non-www) subdomain as source URL.
+  * - affiliated: known to be affiliated in some way. (TBD)
+  * - external: different base domain from source URL.
+  ***/
+
+  var hostType;
+
+  // Case 1: internal
+  if (
+    (www_host == source.hostname) ||
+    (root_host == source.hostname)
+  ) {
+    hostType = "internal";
+
+    if (!data.internal_urls.includes(href))
+      data.internal_urls.push(href);
+
+    // Log www and root requests separately. They are only treated
+    // the same when calculating internal-ness, as they can still have
+    // different technical ramifications.
+    if (!data.internal_domains.includes(url.hostname))
+      data.internal_domains.push(url.hostname);
+  }
+
+  // Case 2: nearby
+  else if (base_host == source_base) {
+    hostType = "nearby";
+
+    if (!data.nearby_urls.includes(href))
+      data.nearby_urls.push(href);
+
+    if (!data.nearby_domains.includes(url.hostname))
+      data.nearby_domains.push(url.hostname);
+  }
+
+  // TODO: Case 3: affiliated
+  // Allow additional provided affiliated suffixes.
+  // For example, allow ".gov" to be considered affiliated.
+
+  // Case 4: external
+  else {
+    hostType = "external";
+
+    if (!data.external_urls.includes(href))
+      data.external_urls.push(href);
+
+    if (!data.external_domains.includes(url.hostname))
+      data.external_domains.push(url.hostname);
+  }
+
+  // Check every URL (even internal/nearby/affiliated ones)
+  // against the list of known services.
+  var known = false;
+  for (name in known_services) {
+    var services = known_services[name];
+    for (service of services) {
+
+      // Either an exact match, or can share a suffix with a known
+      // service-owned hostname.
+      if (
+        (www_host == service) ||
+        (root_host == service) ||
+        (root_host.endsWith(service))
+      ) {
+
+        if (!data.known_services.includes(name))
+          data.known_services.push(name)
+
+        known = true;
+        break;
+      }
+    }
+  }
+
+  // Specifically call out unknown external services for research.
+  if (!known && (hostType == "external")) {
+    if (!data.unknown_services.includes(url.hostname))
+      data.unknown_services.push(url.hostname);
+  }
+};
+
+// For now, a naive base domain calculation.
+// TODO: use the Public Suffix List.
+var baseDomainFor = (input) => {
+  return input.split("\.").slice(-2).join("\.");
+};
```

Comment:
TODO for the future: use the Public Suffix List. (For Lambda, this will probably have to be handled similarly to how it's handled for `pshtt`.)